neo/core/spiketrainlist.py

```diff
@@ -0,0 +1,294 @@
+        elif other and is_spiketrain_or_proxy(other[0]):
+            for obj in other:
+                obj.segment = self.segment
+            if self._items is None:
+                self._spiketrains_from_array()
+            self._items.extend(other)
+            return self
+        else:
+            raise TypeError("Can only add a SpikeTrainList or a list of SpikeTrains in place")
+
+    def __radd__(self, other):
+        """Return other + self"""
+        if isinstance(other, self.__class__):
+            return other._add_spiketrainlists(self)
+        elif other and is_spiketrain_or_proxy(other[0]):
+            for obj in other:
+                obj.segment = self.segment
+            if self._items is None:
+                self._spiketrains_from_array()
+            self._items.extend(other)
+            return self
+        elif len(other) == 0:
+            return self
+        else:
+            if self._items is None:
+                self._spiketrains_from_array()
+            return other + self._items
+
+    def append(self, obj):
+        """L.append(object) -> None -- append object to end"""
+        if not is_spiketrain_or_proxy(obj):
+            raise ValueError("Can only append SpikeTrain objects")
+        if self._items is None:
+            self._spiketrains_from_array()
+        obj.segment = self.segment
+        self._items.append(obj)
+
+    def extend(self, iterable):
+        """L.extend(iterable) -> None -- extend list by appending elements from the iterable"""
+        if self._items is None:
+            self._spiketrains_from_array()
+        for obj in iterable:
+            obj.segment = self.segment
+        self._items.extend(iterable)
+
+    @classmethod
+    def from_spike_time_array(cls, spike_time_array, channel_id_array,
+                              all_channel_ids=None, units=None,
+                              t_start=None, t_stop=None, **annotations):
+        """Create a SpikeTrainList object from an array of spike times
+        and an array of channel ids."""
+        spike_time_array, dim = normalise_times_array(spike_time_array, units)
+        obj = cls()
+        obj._spike_time_array = spike_time_array
+        obj._channel_id_array = channel_id_array
+        obj._all_channel_ids = all_channel_ids
+        obj._spiketrain_metadata = {
+            "t_start": t_start,
+            "t_stop": t_stop
+        }
+        for name, ann_value in annotations.items():
+            if len(ann_value) != len(obj):
+                raise ValueError(f"incorrect length for annotation '{name}'")
+        obj._annotations = annotations
+        return obj
+
+    def _spiketrains_from_array(self):
+        """Convert multiplexed spike time data into a list of SpikeTrain objects"""
+        if self._spike_time_array is None:
+            self._items = []
+        else:
+            if self._all_channel_ids is None:
+                all_channel_ids = np.unique(self._channel_id_array)
```

Comment:
even it is tedious I think we should avoid this magic unique()

Reply:
agreed. done